loader/src/main/java/net/neoforged/fml/loading/mixin/FMLMixinLaunchPlugin.java

```diff
@@ -88,62 +93,124 @@ public void initializeLaunch(ITransformerLoader transformerLoader, NamedPath[] s
-                    for (ModFileParser.MixinConfig potential : file.getMixinConfigs()) {
-                        final var currentInfo = new AnnotationInfo(modId, calculateBehaviorVersion(potential.behaviorVersion()));
-                        var existingInfo = configAnnotationInfo.putIfAbsent(potential.config(), currentInfo);
-                        if (existingInfo != null && !existingInfo.modId().equals(modId)) {
-                            LOGGER.error("Mixin config {} is registered by multiple mods: {} and {}", potential.config(), existingInfo, modId);
-                            ModLoader.addLoadingIssue(ModLoadingIssue.error(
-                                    "fml.modloadingissue.mixin.duplicate_config",
-                                    potential.config(),
-                                    existingInfo,
-                                    modId));
-                        }
-                        if (potential.requiredMods().stream().allMatch(id -> modList.getModFileById(id) != null)) {
-                            Mixins.addConfiguration(potential.config());
-                        } else {
-                            LOGGER.debug("Mixin config {} for mod {} not applied as required mods are missing", potential.config(), modId);
-                        }
-                    }
-                });
-
-        final var configMap = Mixins.getConfigs().stream().collect(
-                Collectors.toMap(Config::getName, Config::getConfig));
-        configAnnotationInfo.forEach((fileName, annotationInfo) -> {
-            if (annotationInfo == null) return;
-
-            final var config = configMap.get(fileName);
+        for (var modFileInfo : modList.getModFiles()) {
+            var modFile = modFileInfo.getFile();
+            for (var mixinConfig : modFile.getMixinConfigs()) {
+                // Validate the mixin version is supported
+                if (!validateMixinBehavior(modFile, mixinConfig)) {
+                    continue;
+                }
+
+                var currentInfo = new AnnotationInfo(modFile, calculateBehaviorVersion(mixinConfig.behaviorVersion()));
+                var existingInfo = configAnnotationInfo.putIfAbsent(mixinConfig.config(), currentInfo);
+                if (existingInfo != null && existingInfo.modFile() != modFile) {
+                    ModLoader.addLoadingIssue(ModLoadingIssue.error(
+                            "fml.modloadingissue.mixin.duplicate_config",
+                            mixinConfig.config(),
+                            existingInfo.modFile)
+                            .withAffectedModFile(modFile));
+                    continue;
+                }
+
+                // Grab the actual mixin config content from the Jar, which also allows us to print a nicer error if it's missing
+                byte[] configContent;
+                try {
+                    configContent = modFile.getContents().readFile(mixinConfig.config());
```

Comment:
I feel like we should do the `requiredMods` check first -- does the file _really_ need to exist if the required mods don't? Maybe not relevant with the existing loader impls; just seems a strange restriction to add.